cache/class-vary-cache.php

```diff
@@ -0,0 +1,164 @@
+	private const VALUE_SEPARATOR = "_--_";
+
+	private static $encryption_enabled = false;
+	private static $groups = [];
+
+	private static $cookie_expiry = MONTH_IN_SECONDS;
+
+	/* nocache */
+	static function set_no_cache_for_user() {
+		self::set_cookie( self::COOKIE_NO_CACHE, 1 );
+	}
+
+	static function remove_no_cache_for_user() {
+		if ( isset( $_COOKIE[ self::COOKIE_NO_CACHE ] ) ) {
+			self::unset_cookie( self::COOKIE_NO_CACHE );
+		}
+	}
+
+	/* Grouping */
+
+	static function register_groups( $groups ) {
+		if( is_array( $groups ) ) {
+			foreach( $groups as $group){
+				self::$groups[ $group ] = '';
+			}
+		} else {
+			self::$groups[ $groups ] = '';
+		}
+
+		self::parseGroupCookie();
+	}
+
+	// will set the group cookie to the added group to indicate Varnish to cache it for those groups
+	static function set_group_for_user( $group, $value ) {
+		//TODO: make sure headers aren't already sent
+		//TODO: only send header if we added or changed things
+		//TODO: don't set the cookie if was already set on the request
+		// validate, process $group, etc.
+		self::$groups[ $group ] = $value;
+		if ( self::is_encryption_enabled() ) {
+
+			self::set_group_cookie_encrypted( self::stringifyGroups() );
+		} else {
+			self::set_group_cookie_plaintext( self::stringifyGroups() );
+		}
+	}
+
+
+	/* check if the user has a group set and optionally that the group value matches */
+	static function is_user_in_group( $group, $value ) {
+		self::parseGroupCookie();
+		if( ! isset ( self::$groups[ $group ] ) ) {
+			return false;
+		}
+
+		return ( null === $value ) || ( self::$groups[ $group ] === $value );
+	}
+
+	static function get_user_groups( ) {
+		self::parseGroupCookie();
+		return self::$groups;
+	}
+
+	static function set_encryption( $value = true ) {
+		static::$encryption_enabled = $value;
+	}
+
+	static function is_encryption_enabled() {
+		return static::$encryption_enabled;
+	}
+
+	static private function set_group_cookie_encrypted( $value, $ttl = null ) {
+		// Validate that we have the secret values
+		if ( ! defined( 'VIP_GO_AUTH_COOKIE_KEY' ) || ! defined( 'VIP_GO_AUTH_COOKIE_IV' ) ) {
+			// TODO: check that values are not empty
+			trigger_error( 'Secrets not defined for encrypted vary cookies', E_USER_WARNING );
+			return;
+		}
+
+		$client_key = constant( 'VIP_GO_AUTH_COOKIE_KEY' );
+		$client_iv = constant( 'VIP_GO_AUTH_COOKIE_IV' );
+		$cookie_value = random_bytpes( 32 ) + '|' . $value . '|' . time() + $ttl;
+		$cipher_cookie = openssl_encrypt( $cookie_value, 'aes-128-cbc', $client_key, 0, $client_iv );
+
+		// TODO: need to scope cookie domain/path + TTL
+		self::set_cookie( self::COOKIE_AUTH, $cipher_cookie );
+	}
+
+	static private function set_group_cookie_plaintext( $value, $ttl = null ) {
+		// TODO: need to scope cookie domain/path + TTL
+		self::set_cookie( self::COOKIE_SEGMENT, $value );
+	}
+
+
+	static private function parseGroupCookie() {
+		if ( isset( $_COOKIE[ self::COOKIE_SEGMENT ] ) ) {
```

Comment:
Needs to check `self::is_encryption_enabled()` and decrypt the cookie before parsing.

Reply:
The decryption has been separated out into a separate method but this still needs a bit more logic to do the right decryption/parsing. Will address in future PR